File: Algorithms/17. Letter Combinations of a Phone Number.rs
use std::collections::HashMap;

impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        
        if digits.len() == 0 { return vec![]; }
        
        // let mut size: usize = 1;
        // digits.chars().for_each(|x| {
        //     if x == '7' || x == '8' {
        //         size *= 4;
        //     }else {
        //         size *= 3;
        //     }
        // });
        // println!("{}", size);
        let mut result: Vec<String> = vec![];
        // println!("{}", digits);
        let letter_map: HashMap<char, Vec<char>> = [
            ('2', vec!['a', 'b', 'c']),
            ('3', vec!['d', 'e', 'f']),
            ('4', vec!['g', 'h', 'i']),
            ('5', vec!['j', 'k', 'l']),
            ('6', vec!['m', 'n', 'o']),
            ('7', vec!['p', 'q', 'r', 's']),
            ('8', vec!['t', 'u', 'v']),
            ('9', vec!['w', 'x', 'y', 'z']),
        ].iter().cloned().collect();
        
        for num in digits.chars() {
            let letters = letter_map.get(&num).unwrap();
            let mut temp: Vec<String> = vec![];

            for c in letters {
                if result.len() == 0 {
                    temp.push(c.to_string());
                } else {
                    for r in result.iter() {
                        temp.push(r.to_owned() + &c.to_string());
                    }
                }
            }
            
            result = temp;
        }
        result
    }
}

File: Algorithms/17. Letter Combinations of a Phone Number.ts
function letterCombinations(digits: string): string[] {
  if (digits.length === 0) {
    return [];
  }

  let result: string[] = [];
  let letterMap: Map<string, string[]> = new Map([
    ["2", ["a", "b", "c"]],
    ["3", ["d", "e", "f"]],
    ["4", ["g", "h", "i"]],
    ["5", ["j", "k", "l"]],
    ["6", ["m", "n", "o"]],
    ["7", ["p", "q", "r", "s"]],
    ["8", ["t", "u", "v"]],
    ["9", ["w", "x", "y", "z"]],
  ]);

  for (const char of digits) {
    const letters = letterMap.get(char);
    const temp: string[] = [];

    for (const letter of letters) {
      if (result.length === 0) {
        temp.push(letter);
      } else {
        for (const r of result) {
          temp.push(`${r}${letter}`);
        }
      }
    }
    result = temp;
  }
  return result;
}
